### ISSUE TITLE
Reduce generic types in `Raft` to one, which would provide all necessary types as associated types

### ISSUE THREAD

**Issue by drmingdrmer** [datafuselabs/openraft#204]
> To be honest, making NodeId a generic type is just what I was planning to do! :DDD

Yeah, but the number of generics then keeps increasing. I was thinking about a different approach: instead of adding generics, reduce them to one, which would provide all necessary types as associated types. Then it's extremely easy to add further required types (one-liner for the user to adapt to the change) and also to set defaults (once the relevant Rust feature is complete; then no effect for the user), so you could even think about providing entire set of runtime features that can be implemented more optimally by the user.

> Thanks for this clean patch!

You are welcome!

_Originally posted by @schreter in https://github.com/datafuselabs/openraft/issues/190#issuecomment-1048510436_

**Comment by schreter** [comment]
I can address this after it's decided how to handle the first step of refactoring (see https://github.com/datafuselabs/openraft/pull/199). Because changing it in parallel will create a huge amount of conflicts :-).

**Comment by schreter** [comment]
Unfortunately, we can't reduce it to one type, but at least the two request/response types are now subsumed in `RaftConfig` and we can add further types on `RaftConfig`.

**Comment by schreter** [comment]
@drmingdrmer BTW, I'm working on the `NodeId` as a parameter, so we can use GUID `NodeId`s. Almost done, I'll post the change tomorrow, after some cleanups. It's quite big, since the config has to be propagated through quite a few structures.

I just noticed, there is no separate issue for it, so I'll refer to this one.

Afterwards, the `Runtime` configuration can be started, so we can provide a deterministic test runtime.

**Comment by drmingdrmer** [comment]
> I just noticed, there is no separate issue for it, so I'll refer to this one.

Issue created!

> Afterwards, the `Runtime` configuration can be started, so we can provide a deterministic test runtime.

😆 😆 😆

**Comment by lichuang** [comment]
@schreter Since you commit to github use your enterprise email(sap.com), which has no related github user, so in the contributors list(https://github.com/datafuselabs/openraft/graphs/contributors) we can not find your name, what is a pity, 

Maybe you can change to a email which has a related github user, so all the `openraft` user can see your commit from the contributors list?

**Comment by schreter** [comment]
@lichuang It didn't occur to me that this is a problem :-). There should be some possibility to link GitHub Enterprise and GitHub accounts, which was requested by my colleagues. If I understand it correctly, that should effectively solve the problem, because then the commits from the SAP address would be linked. Anyway, for now, I'm adding my work mail as a secondary e-mail to GitHub right now, so let's see if it helps.

**Comment by drmingdrmer** [comment]
It works :DDD

https://github.com/datafuselabs/openraft/graphs/contributors

![image](https://user-images.githubusercontent.com/44069/156100549-fc463a21-2ae1-4c23-a6e2-624d05e0be23.png)

**Comment by schreter** [comment]
Type configuration is now implemented, so this issue can be closed.

There is a follow-up in #222

**Comment by drmingdrmer** [comment]
Love your new avatar:D
